### PR TITLE
opentelemetry-cpp: add v1.12.0, drop old versions, update and fix options, fix build requirements

### LINKED ISSUE #18967: [package] opentelemetry-cpp/1.9.1: fails to build due to broken opentelemetry-proto dependency
### Description

run `conan install . --build=missing` with a conanfile:

```
[requires]
opentelemetry-cpp/1.9.1

[generators]
cmake

```
The package installation fails with message:
```
'ERROR: opentelemetry-cpp/1.9.1: required component package 'opentelemetry-proto::' not in dependencies'
The output is observed after 
'opentelemetry-cpp/1.9.1: Package '8865730c41263c8e8733581b5e0ae10b5378ffa9' created' 
```

### Package and Environment Details

* Package Name/Version: **opentelemetry-cpp/1.9.1**
* Operating System+version: **Linux Ubuntu 22.04**

### Conan profile

[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=13
os=Linux

### Steps to reproduce

create conanfile as:

[requires]
opentelemetry-cpp/1.9.1

[generators]
cmake
run conan install . --build=missing

### Logs

<details><summary>Click to expand log</summary>

```
ERROR: opentelemetry-cpp/1.9.1: required component package 'opentelemetry-proto::' not in dependencies
```

</details>
Note: the cmake build passes successfully, and the package seems to be installed, just that this dependency check makes it unusable.

### LINKED ISSUE #20855: [package] opentelemetry-cpp/1.9.1: broken dependency graph
### Description

With the latest recipe of grpc from the 23/10/2023 (0ea683e6b90eb00ac2cdd937700a9e7e), there is a version mismatch in opentelemetry-cpp's dependency graph if both options with_abseil and with_otlp_grpc are set to "True": the direct dep to abseil is with version abseil/20220623.0, and grpc depends on abseil/20230125.3.

### Package and Environment Details

* Package Name/Version: **opentelemetry-cpp/1.8.1**
* Package Name/Version: **opentelemetry-cpp/1.9.1**

### Conan profile

Any

### Steps to reproduce

Build the opentelemetry-cpp package with options with_abseil and with_otlp_grpc set to "True".

### Logs

<details><summary>Click to expand log</summary>

```
ERROR: Version conflict: Conflict between abseil/20230125.3 and abseil/20220623.0 in the graph.
Conflict originates from grpc/1.50.1
```

</details>

### PR DESCRIPTION
**Pull request by valgur**
- Fixes #18967
- Fixes #20855
- Resolves #18647
- Resolves #19759
- Resolves #19793
- Resolves https://github.com/conan-io/conan-center-index/issues/21827
- Resolves https://github.com/conan-io/conan-center-index/issues/21822

`with_otlp` and `with_jaeger` options have been removed in recent versions.

### CODE REVIEW

**Comment by kontrybutor** [comment]
> ### Conan v1 pipeline ✔️
> All green in build 10 (`5a1891488916b64037f1be57bbfcb1deaad16d42`):
> 
> * **opentelemetry-cpp/1.12.0:**
>   All packages built successfully! ([All logs](https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/21332/10/opentelemetry-cpp/1.12.0//summary.json))
> * **opentelemetry-cpp/1.9.1:**
>   All packages built successfully! ([All logs](https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/21332/10/opentelemetry-cpp/1.9.1//summary.json))
> * **opentelemetry-cpp/1.8.3:**
>   All packages built successfully! ([All logs](https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/21332/10/opentelemetry-cpp/1.8.3//summary.json))
> * **opentelemetry-cpp/1.7.0:**
>   All packages built successfully! ([All logs](https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/21332/10/opentelemetry-cpp/1.7.0//summary.json))
> 
> ### Conan v2 pipeline ❌
> > **Note**: Conan v2 builds are now mandatory. Please read our [discussion](https://github.com/conan-io/conan-center-index/discussions/19104) about it.
> 
> The v2 pipeline failed. Please, review the errors and note this is required for pull requests to be merged. In case this recipe is still not ported to Conan 2.x, please, ping `@conan-io/barbarians` on the PR and we will help you.
> 
> See details:
> Failure in build 10 (`5a1891488916b64037f1be57bbfcb1deaad16d42`):
> 
> * **opentelemetry-cpp/1.12.0:**
>   CI failed to create some packages ([All logs](https://c3i.jfrog.io/c3i/misc-v2/summary.html?json=https://c3i.jfrog.io/c3i/misc-v2/logs/pr/21332/10-linux-gcc/opentelemetry-cpp/1.12.0//summary.json))
>   Logs for packageID `5c35ed335c768c900234d0ca09d909522e8303fa`:
>   ```
>   [settings]
>   arch=x86_64
>   build_type=Release
>   compiler=gcc
>   compiler.cppstd=17
>   compiler.libcxx=libstdc++11
>   compiler.version=11
>   os=Linux
>   [options]
>   */*:shared=False
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   ```
>   [...]
>       nlohmann_json/3.11.2#1ded6ae5d200a68ac17c51d528b945e2:da39a3ee5e6b4b0d3255bfef95601890afd80709#2d1a5b1f5d673e1dab536bed20ce000b - Download (conan-center)
>       openssl/3.2.0#d81e0a8c08372368fe7cb757280e2196:5246365bd0fb8fe4176d7e3004664503a288770f#36cc26ad7646e948d80fa8f0ffd2c6bb - Download (conan-center)
>       opentelemetry-cpp/1.12.0#2461ce03102c5e97d7012e9d369b81e4:5c35ed335c768c900234d0ca09d909522e8303fa - Build
>       protobuf/3.21.12#ed2404ae366627005beb2173eab68174:10670e90b03eb1607902ed190ae2fecdc52d69bb#75dba22c6ce1386bc21644af30e1c667 - Download (conan-center)
>       re2/20230301#4482314b6488e5d71e9944f5bedb730d:81d35562505eb9186290237bea6e2684b0ba3cc8#684bce1ba6e0f89a6fd6e6ed9262f99d - Download (conan-center)
>       zlib/1.3#06023034579559bb64357db3a53f88a4:b647c43bfefae3f830561ca202b6cfd935b56205#9f51ace6f715101c7daf0f943c803d69 - Download (conan-center)
>   Build requirements
>       grpc/1.54.3#b9dab567bb1de36e1fababfa2ee5edac:f907664fa3de2578b82af2ad5f031b1feea0c86b - Missing
>       opentelemetry-proto/1.0.0#3e81b1b6e3edfb7ff6f95e72894bdaf8:da39a3ee5e6b4b0d3255bfef95601890afd80709#6651e9db63d50244b2a8286033907aa5 - Download (conan-center)
>       protobuf/3.21.12#ed2404ae366627005beb2173eab68174:10670e90b03eb1607902ed190ae2fecdc52d69bb#75dba22c6ce1386bc21644af30e1c667 - Download (conan-center)
>   Skipped binaries
>       abseil/20230125.3, autoconf/2.71, automake/1.16.5, c-ares/1.19.1, gnu-config/cci.20210814, libtool/2.4.7, m4/1.4.19, meson/1.2.2, ninja/1.11.1, openssl/3.2.0, pkgconf/2.0.3, re2/20230301, zlib/1.3
>   
>   ======== Installing packages ========
>   ERROR: Missing binary: grpc/1.54.3:f907664fa3de2578b82af2ad5f031b1feea0c86b
>   
>   grpc/1.54.3: WARN: Can't find a 'grpc/1.54.3' package binary 'f907664fa3de2578b82af2ad5f031b1feea0c86b' for the configuration:
>   [settings]
>   arch=x86_64
>   build_type=Release
>   compiler=gcc
>   compiler.cppstd=17
>   compiler.libcxx=libstdc++11
>   compiler.version=11
>   os=Linux
>   [options]
>   codegen=True
>   cpp_plugin=True
>   csharp_ext=False
>   csharp_plugin=True
>   fPIC=True
>   node_plugin=True
>   objective_c_plugin=True
>   php_plugin=True
>   python_plugin=True
>   ruby_plugin=True
>   shared=False
>   [requires]
>   abseil/20230125.3.Z
>   c-ares/1.19.Z
>   openssl/3.2.Z
>   protobuf/3.21.Z
>   re2/20230301.0.Z
>   zlib/1.3.Z
>   
>   ERROR: Missing prebuilt package for 'grpc/1.54.3'
>   Check the available packages using 'conan list grpc/1.54.3:* -r=remote'
>   or try to build locally from sources using the '--build=grpc/1.54.3' argument
>   
>   More Info at 'https://docs.conan.io/2/knowledge/faq.html#error-missing-prebuilt-package'
>   ```
> * **opentelemetry-cpp/1.9.1:**
>   Didn't run or was cancelled before finishing
> * **opentelemetry-cpp/1.8.3:**
>   Didn't run or was cancelled before finishing
> * **opentelemetry-cpp/1.7.0:**
>   Didn't run or was cancelled before finishing
> 
> **Note:** To save resources, CI tries to finish as soon as an error is found. For this reason you might find that not all the references have been launched or not all the configurations for a given reference. Also, take into account that we cannot guarantee the order of execution as it depends on CI workload and workers availability.

Do we know why Conan cannot find the `grpc/1.54.3` binary? Looks like it has been added to CCI already: https://conan.io/center/recipes/grpc?version=1.54.3

**Comment by kontrybutor** [comment]
> (...)
> 
> ======== Installing packages ========
> ERROR: Missing binary: grpc/1.54.3:c6ae82072811d68596ca608e40565b276bf7ec46
> 
> grpc/1.54.3: WARN: Can't find a 'grpc/1.54.3' package binary 'c6ae82072811d68596ca608e40565b276bf7ec46' for the configuration:
> [settings]
> arch=armv8
> build_type=Release
> compiler=apple-clang
> compiler.cppstd=17
> compiler.libcxx=libc++
> compiler.version=13
> os=Macos
> [options]
> codegen=True
> cpp_plugin=True
> csharp_ext=False
> csharp_plugin=True
> fPIC=True
> node_plugin=True
> objective_c_plugin=True
> php_plugin=True
> python_plugin=True
> ruby_plugin=True
> shared=False
> [requires]
> abseil/20230125.3.Z
> c-ares/1.19.Z
> openssl/3.2.Z
> protobuf/3.21.Z
> re2/20230301.0.Z
> zlib/1.3.Z
> 
> ERROR: Missing prebuilt package for 'grpc/1.54.3'
> Check the available packages using 'conan list grpc/1.54.3:* -r=remote'
> or try to build locally from sources using the '--build=grpc/1.54.3' argument
> 
> More Info at 'https://docs.conan.io/2/knowledge/faq.html#error-missing-prebuilt-package'
> 


Could someone explain me why CI cannot find the prebuild `grpc` binary on CCI? There is a packageID `d299b4695808861cdd46622bf1ec6d0007301bee` that has the same settings and options as in the quoted log.
https://conan.io/center/recipes/grpc?version=1.54.3

File: recipes/opentelemetry-cpp/all/conanfile.py [hunk 1]
```diff
@@ -67,24 +66,37 @@ class OpenTelemetryCppConan(ConanFile):
...
 
     def config_options(self):
         if self.settings.os == "Windows":
             self.options.rm_safe("fPIC")
+        if Version(self.version) >= "1.10":
+            del self.options.with_jaeger
 
     def configure(self):
         if self.options.shared:
             self.options.rm_safe("fPIC")
```
Comment (uilianries) [inline recipes/opentelemetry-cpp/all/conanfile.py:99]:
```suggestion
    if self.options.with_otlp != "deprecated":
        self.output.warning(f"{self.ref}:with_otlp option is deprecated, do not use anymore. Please, consider with_otlp_grpc or with_otlp_http instead.")
```

https://github.com/conan-io/conan-center-index/blob/f174071b1e873360c644b401e61a66bd3c8f9aac/docs/faqs.md#can-i-remove-an-option-from-a-recipe

File: recipes/opentelemetry-cpp/all/conanfile.py [hunk 2]
```diff
@@ -93,60 +105,51 @@ def requirements(self):
...
-                self.requires("opentelemetry-proto/0.11.0")
-            else:
-                self.requires("opentelemetry-proto/0.20.0")
-
-            if self.options.get_safe("with_otlp_grpc"):
-                self.requires("grpc/1.50.1")
+        if self.options.with_otlp or self.options.with_otlp_grpc:
+            self.requires("protobuf/3.21.12", transitive_headers=True, transitive_libs=True)
+            if self.options.with_otlp_grpc:
+                self.requires("grpc/1.54.3", transitive_headers=True, transitive_libs=True)
```
Comment (uilianries) [inline recipes/opentelemetry-cpp/all/conanfile.py:113]:
```suggestion
        if self.options.with_otlp_grpc:
            self.requires("protobuf/3.21.12", transitive_headers=True, transitive_libs=True)
            self.requires("grpc/1.54.3", transitive_headers=True, transitive_libs=True)
```

Please, do not consider a deprecated option as part of recipe conditions.

File: recipes/opentelemetry-cpp/all/conanfile.py [hunk 3]
```diff
@@ -159,11 +162,17 @@ def validate(self):
...
     def build_requirements(self):
-        if self.options.with_otlp:
-            self.tool_requires("protobuf/3.21.9")
-            if self.options.get_safe("with_otlp_grpc"):
-                self.tool_requires("grpc/1.50.1")
+        if self.options.with_otlp or self.options.with_otlp_grpc:
+            self.tool_requires("opentelemetry-proto/1.0.0")
+            self.tool_requires("protobuf/<host_version>")
+            if self.options.with_otlp_grpc:
+                self.tool_requires("grpc/<host_version>")
```
Comment (uilianries) [inline recipes/opentelemetry-cpp/all/conanfile.py:175]:
```suggestion
        if self.options.with_otlp_grpc:
            self.tool_requires("opentelemetry-proto/1.0.0")
            self.tool_requires("protobuf/<host_version>")
            self.tool_requires("grpc/<host_version>")
```
Please, do not include deprecated options as part of the condition flow.

File: recipes/opentelemetry-cpp/all/conanfile.py [hunk 4]
```diff
@@ -176,72 +185,68 @@ def _create_cmake_module_variables(self, module_file):
...
+        tc = CMakeToolchain(self)
+        tc.cache_variables["BUILD_TESTING"] = False
+        tc.cache_variables["BUILD_BENCHMARK"] = False
+        tc.cache_variables["WITH_EXAMPLES"] = False
+        tc.cache_variables["WITH_NO_DEPRECATED_CODE"] = self.options.with_no_deprecated_code
+        tc.cache_variables["WITH_STL"] = self.options.with_stl
+        tc.cache_variables["WITH_GSL"] = self.options.with_gsl
+        tc.cache_variables["WITH_ABSEIL"] = self.options.with_abseil
+        if Version(self.version) < "1.10":
+            tc.cache_variables["WITH_OTLP"] = self.options.with_otlp
```
Comment (uilianries) [inline recipes/opentelemetry-cpp/all/conanfile.py:208]:
```suggestion
            tc.cache_variables["WITH_OTLP"] = self.options.with_otlp_grpc or self.options.with_otlp_http
```

**Review by uilianries — COMMENTED** [review]
After marking an option as deprecated, please, do not keep maintaining it as part of the recipe.

**Review by franramirez688 — APPROVED** [review]
LGTM! 👏 

@uilianries do you want to take another glance?

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**